File: 15/numbers.py
data = (10,16,6,0,1,17)

last_spoken = {v: k+1 for k, v in enumerate(data[:-1])}

prev = data[-1]
turn = len(data)

while turn < 30000000:
    if prev not in last_spoken:
        new = 0
    else:
        new = turn - last_spoken[prev]

    last_spoken[prev] = turn

    prev = new
    turn += 1

print(turn, prev)

File: 15/numbers.go
package main

import (
	"fmt"
)

func main() {
	data := []int{10, 16, 6, 0, 1, 17}
	lastSpoken := map[int]int{}
	for k, v := range data {
		if k < len(data)-1 {
			lastSpoken[v] = k + 1
		}
	}
	prev := data[len(data)-1]
	turn := len(data)

	var next int
	for turn < 30_000_000 {
		prevTurn, ok := lastSpoken[prev]
		if ok {
			next = turn - prevTurn
		} else {
			next = 0
		}
		lastSpoken[prev] = turn
		prev = next
		turn++
	}

	fmt.Println(turn, prev)
}
